sfc_climo_gen/man/man3/source_grid.3
.TH "source_grid" 3 "Tue Jan 21 2025" "Version 1.13.0" "sfc_climo_gen" \" -*- nroff -*-
.ad l
.nh
.SH NAME
source_grid \- Read grid specs, date information and land/sea mask for the source data that will be interpolated to the model grid\&.  

.SH SYNOPSIS
.br
.PP
.SS "Functions/Subroutines"

.in +1c
.ti -1c
.RI "subroutine, public \fBdefine_source_grid\fP (localpet, npets, input_file)"
.br
.RI "Defines esmf grid object for source grid\&. "
.ti -1c
.RI "subroutine, public \fBsource_grid_cleanup\fP"
.br
.RI "Free up memory associated with this module\&. "
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "integer, dimension(:), allocatable, public \fBday_of_rec\fP"
.br
.RI "Day of each time record with respect to Jan 1\&. "
.ti -1c
.RI "character(len=50), dimension(:), allocatable, public \fBfield_names\fP"
.br
.RI "Names of fields to be processed\&. "
.ti -1c
.RI "type(esmf_grid), public \fBgrid_src\fP"
.br
.RI "ESMF grid object for the source grid\&. "
.ti -1c
.RI "integer, public \fBi_src\fP"
.br
.RI "i dimension of the source grid\&. "
.ti -1c
.RI "integer, public \fBj_src\fP"
.br
.RI "j dimension of the source grid\&. "
.ti -1c
.RI "integer, public \fBnum_fields\fP"
.br
.RI "Number of fields in the file\&. "
.ti -1c
.RI "integer, public \fBnum_records\fP"
.br
.RI "Number of fields times time records\&. "
.ti -1c
.RI "integer, public \fBnum_time_recs\fP"
.br
.RI "Number of time records\&. "
.ti -1c
.RI "character(len=75), public \fBsource\fP"
.br
.RI "Original source of the data\&. "
.in -1c
.SH "Detailed Description"
.PP 
Read grid specs, date information and land/sea mask for the source data that will be interpolated to the model grid\&. 

Also, sets up the ESMF grid object for the source grid\&. Source grid is assumed to be global lat/lon\&.
.PP
\fBAuthor:\fP
.RS 4
George Gayno 
.RE
.PP
\fBDate:\fP
.RS 4
2018 
.RE
.PP

.SH "Function/Subroutine Documentation"
.PP 
.SS "subroutine, public source_grid::define_source_grid (integer, intent(in) localpet, integer, intent(in) npets, character(len=*), intent(in) input_file)"

.PP
Defines esmf grid object for source grid\&. Retrieves date and field information from source file\&.
.PP
Read date information from input source data file\&. Create esmf grid object for the source grid\&.
.PP
\fBParameters:\fP
.RS 4
\fIlocalpet\fP mpi task number 
.br
\fInpets\fP total number mpi tasks 
.br
\fIinput_file\fP file containing the source grid data\&. 
.RE
.PP
\fBAuthor:\fP
.RS 4
George Gayno 
.RE
.PP
\fBDate:\fP
.RS 4
2018 
.RE
.PP

.PP
Definition at line 52 of file source_grid\&.F90\&.
.PP
References day_of_rec, utils::error_handler(), field_names, grid_src, i_src, j_src, utils::netcdf_err(), num_fields, num_records, num_time_recs, and source\&.
.PP
Referenced by driver()\&.
.SS "subroutine, public source_grid::source_grid_cleanup ()"

.PP
Free up memory associated with this module\&. 
.PP
\fBAuthor:\fP
.RS 4
George Gayno 
.RE
.PP
\fBDate:\fP
.RS 4
2018 
.RE
.PP

.PP
Definition at line 404 of file source_grid\&.F90\&.
.PP
References day_of_rec, field_names, and grid_src\&.
.PP
Referenced by driver()\&.
.SH "Variable Documentation"
.PP 
.SS "integer, dimension(:), allocatable, public source_grid::day_of_rec"

.PP
Day of each time record with respect to Jan 1\&. 
.PP
Definition at line 31 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), output(), source_grid_cleanup(), and output_frac_cats::writeit()\&.
.SS "character(len=50), dimension(:), allocatable, public source_grid::field_names"

.PP
Names of fields to be processed\&. 
.PP
Definition at line 21 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), interp_frac_cats(), output(), output_frac_cats::output_driver(), source_grid_cleanup(), and output_frac_cats::writeit()\&.
.SS "type(esmf_grid), public source_grid::grid_src"

.PP
ESMF grid object for the source grid\&. 
.PP
Definition at line 34 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), interp_frac_cats(), and source_grid_cleanup()\&.
.SS "integer, public source_grid::i_src"

.PP
i dimension of the source grid\&. 
.PP
Definition at line 24 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), and interp_frac_cats()\&.
.SS "integer, public source_grid::j_src"

.PP
j dimension of the source grid\&. 
.PP
Definition at line 25 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), and interp_frac_cats()\&.
.SS "integer, public source_grid::num_fields"

.PP
Number of fields in the file\&. Some files have more than one (ex: the 4-component albedo)\&. 
.PP
Definition at line 28 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), and output()\&.
.SS "integer, public source_grid::num_records"

.PP
Number of fields times time records\&. 
.PP
Definition at line 26 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), and output()\&.
.SS "integer, public source_grid::num_time_recs"

.PP
Number of time records\&. 
.PP
Definition at line 27 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), interp(), output(), and output_frac_cats::writeit()\&.
.SS "character(len=75), public source_grid::source"

.PP
Original source of the data\&. 
.PP
Definition at line 22 of file source_grid\&.F90\&.
.PP
Referenced by define_source_grid(), output(), and output_frac_cats::writeit()\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for sfc_climo_gen from the source code\&.


sorc/sfc_climo_gen.fd/source_grid.F90
!> @file
!! @brief Read grid specs, date information and land/sea mask for
!! the source data that will be interpolated to the model grid.
!! @author George Gayno @date 2018

!> Read grid specs, date information and land/sea mask for
!! the source data that will be interpolated to the model grid.
!! Also, sets up the ESMF grid object for the source grid.
!! Source grid is assumed to be global lat/lon.
!!
!! @author George Gayno @date 2018
module source_grid

 use esmf
 use utils

 implicit none

 private 

 character(len=50), allocatable, public :: field_names(:) !< Names of fields to be processed.
 character(len=75), public              :: source !< Original source of the data.

 integer, public               :: i_src !< i dimension of the source grid.
 integer, public               :: j_src !< j dimension of the source grid.
 integer, public               :: num_records !< Number of fields times time records.
 integer, public               :: num_time_recs !< Number of time records.
 integer, public               :: num_fields !< Number of fields in the file. Some
                                             !! files have more than one (ex: 
                                             !! the 4-component albedo).
 integer, allocatable, public  :: day_of_rec(:) !< Day of each time record with
                                                !! respect to Jan 1.

 type(esmf_grid), public       :: grid_src !< ESMF grid object for the source grid.

 public                        :: define_source_grid
 public                        :: source_grid_cleanup

 contains

 !> Defines esmf grid object for source grid. Retrieves date and field
 !! information from source file.
 !!
 !! Read date information from input source data file.
 !! Create esmf grid object for the source grid.
 !!
 !! @param[in] localpet mpi task number
 !! @param[in] npets total number mpi tasks
 !! @param[in] input_file file containing the source grid data.
 !! @author George Gayno @date 2018
 subroutine define_source_grid(localpet, npets, input_file)

 use mpi
 use netcdf

 implicit none

 character(len=*), intent(in)     :: input_file

 integer, intent(in)              :: localpet, npets
 
 character(len=50)                :: field_names_save(100)
 character(len=50)                :: vname

 integer                          :: dimid, dims(1), ncid, status
 integer                          :: count, num_vars, n
 integer                          :: rc, varid, i, j, i_src_corner
 integer(esmf_kind_i4), pointer   :: mask_ptr(:,:)
 integer                          :: clb(2), cub(2)
 integer                          :: clb_corner(2), cub_corner(2)

 real(esmf_kind_r4), allocatable  :: mask_global(:,:)
 real(esmf_kind_r8), allocatable  :: lat_global(:)
 real(esmf_kind_r8), allocatable  :: lon_global(:)
 real(esmf_kind_r8), allocatable  :: lat_corner_global(:)
 real(esmf_kind_r8), allocatable  :: lon_corner_global(:)
 real(esmf_kind_r4), pointer      :: mask_field_ptr(:,:)
 real(esmf_kind_r8), pointer      :: lat_ptr(:,:)
 real(esmf_kind_r8), pointer      :: lon_ptr(:,:)
 real(esmf_kind_r8), pointer      :: lat_corner_ptr(:,:)
 real(esmf_kind_r8), pointer      :: lon_corner_ptr(:,:)
 real                             :: lon_extent
 real(esmf_kind_r4)               :: missing

 type(esmf_field)                 :: mask_field
 type(esmf_polekind_flag)         :: polekindflag(2)

 print*,"- OPEN FILE: ", trim(input_file)
 status = nf90_open(input_file, nf90_nowrite, ncid)
 call netcdf_err(status, "OPENING INPUT SOURCE FILE")

 status = nf90_get_att(ncid, nf90_global, 'source', source)
 if (status /= nf90_noerr) source="unknown"

 if(localpet == 0) print*,'- SOURCE OF DATA IS: ', trim(source)

 status = nf90_inq_dimid(ncid, 'idim', dimid)
 call netcdf_err(status, "READING I DIMENSION ID OF SOURCE DATA")
 status = nf90_inquire_dimension(ncid, dimid, len=i_src)
 call netcdf_err(status, "READING I DIMENSION OF SOURCE DATA")

 status = nf90_inq_dimid(ncid, 'jdim', dimid)
 call netcdf_err(status, "READING J DIMENSION ID OF SOURCE DATA")
 status = nf90_inquire_dimension(ncid, dimid, len=j_src)
 call netcdf_err(status, "READING J DIMENSION OF SOURCE DATA")

 if(localpet == 0) print*,'- I/J DIMENSION OF SOURCE DATA: ', i_src, j_src

 allocate(lat_global(j_src))
 status = nf90_inq_varid(ncid, 'lat', varid)
 call netcdf_err(status, "READING SOURCE DATA LATITUDE ID")
 status = nf90_get_var(ncid, varid, lat_global)
 call netcdf_err(status, "READING SOURCE DATA LATITUDES")

 allocate(lon_global(i_src))
 status = nf90_inq_varid(ncid, 'lon', varid)
 call netcdf_err(status, "READING SOURCE DATA LONGITUDE ID")
 status = nf90_get_var(ncid, varid, lon_global)
 call netcdf_err(status, "READING SOURCE DATA LONGITUDE")

 allocate(lat_corner_global(j_src+1))
 status = nf90_inq_varid(ncid, 'lat_corner', varid)
 call netcdf_err(status, "READING SOURCE DATA CORNER LATITUDE ID")
 status = nf90_get_var(ncid, varid, lat_corner_global)
 call netcdf_err(status, "READING SOURCE DATA CORNER LATITUDE")

!-----------------------------------------------------------------------
! Dimension of lon_corner depends on whether input data is periodic
! (global) or regional.
!-----------------------------------------------------------------------

 status = nf90_inq_varid(ncid, 'lon_corner', varid)
 call netcdf_err(status, "READING SOURCE DATA CORNER LONGITUDE ID")
 status = nf90_inquire_variable(ncid, varid, dimids=dims)
 call netcdf_err(status, "READING SOURCE DATA CORNER LONGITUDE DIMIDS")
 status = nf90_inquire_dimension(ncid, dims(1), len=i_src_corner)
 call netcdf_err(status, "READING SOURCE DATA CORNER LONGITUDE DIMS")
 allocate(lon_corner_global(i_src_corner))
 status = nf90_get_var(ncid, varid, lon_corner_global)
 call netcdf_err(status, "READING SOURCE DATA CORNER LONGITUDE")

 status = nf90_inq_dimid(ncid, 'time', dimid)
 call netcdf_err(status, "READING SOURCE DATA TIME ID")
 status = nf90_inquire_dimension(ncid, dimid, len=num_time_recs)
 call netcdf_err(status, "READING SOURCE DATA NUM TIME PERIODS")

 allocate(day_of_rec(num_time_recs))
 status = nf90_inq_varid(ncid, 'time', varid)
 call netcdf_err(status, "READING SOURCE DATA TIME VARID")
 status = nf90_get_var(ncid, varid, day_of_rec)
 call netcdf_err(status, "READING SOURCE DATA DAY OF RECORD")

 print*,'- SOURCE DATA DAYS OF RECORD: ', day_of_rec

 status = nf90_inquire(ncid, nVariables=num_vars)
 call netcdf_err(status, "READING NUMBER OF VARIABLES SOURCE DATA")

!-----------------------------------------------------------------------
! Assumes files contain records for time, lat, lon, lat_corner, 
! lon_corner.  So number of fields processed will be the total
! number of variables minus 5.  
!-----------------------------------------------------------------------

! NOTE: the new BNU soil type data contains extra records for
! sand and clay percentages. These extra records are not need yet,
! so add logic to temporarily ignore them.

 count = 0
 do n = 1, num_vars

  status = nf90_inquire_variable(ncid, n, name=vname)
  call netcdf_err(status, "READING FIELD NAMES")

  if (trim(vname) == 'time') cycle
  if (trim(vname) == 'lon') cycle
  if (trim(vname) == 'lon_corner') cycle
  if (trim(vname) == 'lat') cycle
  if (trim(vname) == 'lat_corner') cycle
  if (trim(vname) == 'clay_lev1') cycle
  if (trim(vname) == 'clay_top') cycle
  if (trim(vname) == 'sand_lev1') cycle
  if (trim(vname) == 'sand_top') cycle

  count = count + 1
  field_names_save(count) = vname

 enddo

 num_fields = count
 num_records = num_vars * num_time_recs

 allocate(field_names(num_fields))

 field_names = field_names_save(1:num_fields)

 if(localpet==0) print*,'- FIELDS TO BE PROCESSED: ', field_names

 if (localpet == 0) then
   allocate(mask_global(i_src,j_src))
   status = nf90_inq_varid(ncid, field_names(1), varid)
   call netcdf_err(status, "READING FIELD ID")
   status = nf90_get_var(ncid, varid, mask_global)
   call netcdf_err(status, "READING FIELD")
 else
   allocate(mask_global(0,0))
 endif

!--------------------------------------------------------------------------
! Read in missing value.  This is used to mask out data at non-land
! points.
!--------------------------------------------------------------------------

 status = nf90_inq_varid(ncid, field_names(1), varid)
 call netcdf_err(status, "READING FIELD 1 ID")
 status=nf90_get_att(ncid, varid, 'missing_value', missing)
 call netcdf_err(status, "READING MISSING VALUE")
 status = nf90_close(ncid)




!--------------------------------------------------------------------------
! Create ESMF grid object for the source data grid.  Check if 
! data is periodic in the east/west direction.
!
! Note: When using regional data, there is always the chance of
! the target grid being located outside the input grid.  ESMF
! support recommends passing back the dstField (esmf_typekind_i4) from
! all calls to FieldRegridStore and checking for the
! ESMF_REGRIDSTATUS_OUTSIDE flag.
!--------------------------------------------------------------------------

 lon_extent = mod((lon_global(i_src)-lon_global(1))-1+3600,360.)+1.0
 
 if (lon_extent < 350.0) then

   print*,"- CALL GridCreateNoPeriDim FOR SOURCE GRID"
   grid_src = ESMF_GridCreateNoPeriDim(minIndex=(/1,1/), &
                                    maxIndex=(/i_src,j_src/), &
                                    coordSys=ESMF_COORDSYS_SPH_DEG, &
                                    regDecomp=(/1,npets/),  &
                                    name="source_grid", &
                                    indexflag=ESMF_INDEX_GLOBAL, rc=rc)
   if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
      call error_handler("IN GridCreateNoPeriDim.", rc)

 else

   polekindflag(1:2) = ESMF_POLEKIND_MONOPOLE

   print*,"- CALL GridCreate1PeriDim FOR SOURCE GRID"
   grid_src = ESMF_GridCreate1PeriDim(minIndex=(/1,1/), &
                                    maxIndex=(/i_src,j_src/), &
                                    polekindflag=polekindflag, &
                                    periodicDim=1, &
                                    poleDim=2,  &
                                    coordSys=ESMF_COORDSYS_SPH_DEG, &
                                    regDecomp=(/1,npets/),  &
                                    name="source_grid", &
                                    indexflag=ESMF_INDEX_GLOBAL, rc=rc)
   if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
      call error_handler("IN GridCreate1PeriDim.", rc)

 endif

 print*,"- CALL FieldCreate FOR SOURCE GRID LANDMASK."
 mask_field = ESMF_FieldCreate(grid_src, &
                               typekind=ESMF_TYPEKIND_R4, &
                               staggerloc=ESMF_STAGGERLOC_CENTER, &
                               name="source grid land mask", &
                               rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN FieldCreate.", rc)

 print*,"- CALL FieldScatter FOR SOURCE GRID MASK."
 call ESMF_FieldScatter(mask_field, mask_global, rootpet=0, rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN FieldScatter.", rc)

 print*,"- CALL GridAddItem FOR SOURCE GRID MASK."
 call ESMF_GridAddItem(grid_src, &
                       itemflag=ESMF_GRIDITEM_MASK, &
                       staggerloc=ESMF_STAGGERLOC_CENTER, &
                       rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridAddItem.", rc)

 print*,"- CALL GridGetItem FOR SOURCE GRID MASK."
 nullify(mask_ptr)
 call ESMF_GridGetItem(grid_src, &
                       itemflag=ESMF_GRIDITEM_MASK, &
                       farrayPtr=mask_ptr, &
                       totalLBound=clb,  &
                       totalUBound=cub, &
                       rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridGetItem", rc)
 
 print*,"- CALL FieldGet FOR SOURCE GRID LANDMASK."
 nullify(mask_field_ptr)
 call ESMF_FieldGet(mask_field, &
                    farrayPtr=mask_field_ptr,  &
                    rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN FieldGet.", rc)

 do j = clb(2), cub(2)
   do i = clb(1), cub(1)
     if ( abs(mask_field_ptr(i,j)-missing) < 0.001) then
       mask_ptr(i,j) = 0
     else
       mask_ptr(i,j) = 1
     endif
   enddo
 enddo

 deallocate(mask_global)

 print*,"- CALL FieldDestroy FOR SOURCE GRID LAND MASK."
 call ESMF_FieldDestroy(mask_field,rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN FieldDestroy.", rc)

! Set lat/lons of grid points

 print*,"- CALL GridAddCoord FOR SOURCE GRID CENTER LOCATION."
 call ESMF_GridAddCoord(grid_src, &
                        staggerloc=ESMF_STAGGERLOC_CENTER, rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridAddCoord.", rc)

 print*,"- CALL GridGetCoord FOR SOURCE GRID CENTER LONGITUDE."
 nullify(lon_ptr)
 call ESMF_GridGetCoord(grid_src, &
                        staggerLoc=ESMF_STAGGERLOC_CENTER, &
                        coordDim=1, &
                        farrayPtr=lon_ptr, rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridGetCoord.", rc)

 print*,"- CALL GridGetCoord FOR SOURCE GRID CENTER LATITUDE."
 nullify(lat_ptr)
 call ESMF_GridGetCoord(grid_src, &
                        staggerLoc=ESMF_STAGGERLOC_CENTER, &
                        coordDim=2, &
                        farrayPtr=lat_ptr, rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridGetCoord.", rc)

 do j = clb(2), cub(2)
   lat_ptr(:,j) = lat_global(j)
 enddo

 do i = clb(1), cub(1)
   lon_ptr(i,:) = lon_global(i)
 enddo

 print*,"- CALL GridAddCoord FOR SOURCE GRID CORNER LOCATION."
 call ESMF_GridAddCoord(grid_src, &
                        staggerloc=ESMF_STAGGERLOC_CORNER, &
                        rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridAddCoord.", rc)

 print*,"- CALL GridGetCoord FOR SOURCE GRID CORNER LONGITUDE."
 nullify(lon_corner_ptr)
 call ESMF_GridGetCoord(grid_src, &
                        staggerLoc=ESMF_STAGGERLOC_CORNER, &
                        coordDim=1, &
                        farrayPtr=lon_corner_ptr, &
                        rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridGetCoord.", rc)

 print*,"- CALL GridGetCoord FOR SOURCE GRID CORNER LATITUDE."
 nullify(lat_corner_ptr)
 call ESMF_GridGetCoord(grid_src, &
                        staggerLoc=ESMF_STAGGERLOC_CORNER, &
                        coordDim=2, &
                        computationalLBound=clb_corner, &
                        computationalUBound=cub_corner, &
                        farrayPtr=lat_corner_ptr, &
                        rc=rc)
 if(ESMF_logFoundError(rcToCheck=rc,msg=ESMF_LOGERR_PASSTHRU,line=__LINE__,file=__FILE__)) &
    call error_handler("IN GridGetCoord.", rc)

 do j = clb_corner(2), cub_corner(2)
   lat_corner_ptr(:,j) = lat_corner_global(j)
 enddo

 do i = clb_corner(1), cub_corner(1)
   lon_corner_ptr(i,:) = lon_corner_global(i)
 enddo

 deallocate(lat_global)
 deallocate(lon_global)
 deallocate(lat_corner_global)
 deallocate(lon_corner_global)

 end subroutine define_source_grid

 !> Free up memory associated with this module.
 !!
 !! @author George Gayno @date 2018 
 subroutine source_grid_cleanup

 implicit none

 integer  :: rc

 print*,"- CALL GridDestroy FOR SOURCE GRID."
 call ESMF_GridDestroy(grid_src,rc=rc)

 deallocate(day_of_rec)
 deallocate(field_names)

 end subroutine source_grid_cleanup

 end module source_grid
